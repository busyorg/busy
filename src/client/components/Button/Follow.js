import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { injectIntl } from 'react-intl';
import { Icon } from 'antd';
import './Follow.less';

export class FollowPure extends React.Component {
  static propTypes = {
    intl: PropTypes.shape().isRequired,
    isFollowed: PropTypes.bool,
    pending: PropTypes.bool,
    onClick: PropTypes.func,
  };

  static defaultProps = {
    isFollowed: false,
    pending: false,
    onClick: () => {},
  };

  constructor(props) {
    super(props);
    this.state = {
      isHovered: false,
    };
  }

  onMouseOver = () => this.setState({ isHovered: true });

  onMouseOut = () => this.setState({ isHovered: false });

  handleClick = (e) => {
    e.preventDefault();
    if (this.props.pending) return;
    this.props.onClick(e);
  };

  render() {
    const { intl, isFollowed, pending } = this.props;
    const { isHovered } = this.state;

    let followingText = intl.formatMessage({ id: 'follow', defaultMessage: 'Follow' });
    if (isFollowed && !(isHovered || pending)) {
      followingText = intl.formatMessage({ id: 'followed', defaultMessage: 'Following' });
    } else if (isFollowed && isHovered && !pending) {
      followingText = intl.formatMessage({ id: 'unfollow', defaultMessage: 'Unfollow' });
    } else if (isFollowed && pending) {
      followingText = intl.formatMessage({ id: 'unfollowing', defaultMessage: 'Unfollowing' });
    } else if (!isFollowed && isHovered && !pending) {
      followingText = intl.formatMessage({ id: 'follow', defaultMessage: 'Follow' });
    } else if (!isFollowed && pending) {
      followingText = intl.formatMessage({ id: 'followed', defaultMessage: 'Following' });
    }

    return (
      <button
        className={classNames('Follow', {
          'Follow--danger': isFollowed && (isHovered || pending),
        })}
        onClick={this.handleClick}
        onMouseOver={this.onMouseOver}
        onMouseOut={this.onMouseOut}
      >
        {pending && <Icon type="loading" />}
        {followingText}
      </button>
    );
  }
}

export default injectIntl(FollowPure);
