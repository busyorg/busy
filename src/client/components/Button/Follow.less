@import (reference) '../../styles/custom.less';
@import (reference) '../../styles/mixins.less';

.Follow {
  .btn;
  .btn-primary;

  .colorize-button(@purple, #2a3574);

  color: @white-darker;

  & > .anticon {
    margin-right: 12px;
  }
}

.Follow--secondary {
  color: @purple !important;
  background-color: @blue-hawkesBlue !important;
  border: 1px solid @purple;

  &:hover {
    background: darken(@blue-hawkesBlue, 15%, 'relative') !important;
    border-color: darken(@purple, 15%, 'relative') !important;
  }
}

.Follow--danger {
  color: @white !important;
  .colorize-button(@red, @red);
}
