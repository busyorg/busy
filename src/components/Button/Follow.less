@import "../../styles/custom.less";
@import "../../styles/mixins.less";

.Follow {
  .btn;
  .btn-primary;

  .colorize-button(@purple, #2a3574);

  color: @white-darker;

  & > .anticon {
    margin-right: 12px;
  }
}

.Follow--danger {
  color: #fff !important;
  .colorize-button(@red, @red);
}
