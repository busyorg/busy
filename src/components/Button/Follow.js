import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Icon } from 'antd';
import './Follow.less';

class Follow extends React.Component {
  static propTypes = {
    isFollowed: PropTypes.bool,
    pending: PropTypes.bool,
    onClick: PropTypes.func,
  };

  static defaultProps = {
    isFollowed: false,
    pending: false,
    onClick: () => {},
  };

  constructor(props) {
    super(props);
    this.state = {
      isHovered: false,
    };
  }

  onMouseOver = () => this.setState({ isHovered: true });

  onMouseOut = () => this.setState({ isHovered: false });

  handleClick = (e) => {
    e.preventDefault();
    if (this.props.pending) return;
    this.props.onClick(e);
  };

  render() {
    const { isFollowed, pending } = this.props;
    const { isHovered } = this.state;

    let followingText = 'Follow'; // default text
    if (isFollowed && !(isHovered || pending)) {
      followingText = 'Followed';
    } else if (isFollowed && isHovered && !pending) {
      followingText = 'Unfollow';
    } else if (isFollowed && pending) {
      followingText = 'Unfollowing';
    } else if (!isFollowed && isHovered && !pending) {
      followingText = 'Follow';
    } else if (!isFollowed && pending) {
      followingText = 'Following';
    }

    return (
      <button
        className={classNames('Follow', {
          'Follow--danger': isFollowed && (isHovered || pending),
        })}
        onClick={this.handleClick}
        onMouseOver={this.onMouseOver}
        onMouseOut={this.onMouseOut}
      >
        {pending && <Icon type="loading" />}
        {followingText}
      </button>
    );
  }
}

export default Follow;
