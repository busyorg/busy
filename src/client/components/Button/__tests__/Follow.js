import React from 'react';
import { shallow } from 'enzyme';
import { IntlProvider } from 'react-intl';
import { FollowPure as Follow } from '../Follow';

describe('<Follow />', () => {
  it('renders without exploding', () => {
    const intlProvider = new IntlProvider({ locale: 'en' }, {});
    const { intl } = intlProvider.getChildContext();
    const props = {
      intl,
      isFollowed: true,
      pending: false,
    };
    const wrapper = shallow(<Follow {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
