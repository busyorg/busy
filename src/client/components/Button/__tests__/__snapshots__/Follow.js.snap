// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Follow /> renders without exploding 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FollowPure
    intl={
      Object {
        "defaultFormats": Object {},
        "defaultLocale": "en",
        "formatDate": [Function],
        "formatHTMLMessage": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatPlural": [Function],
        "formatRelative": [Function],
        "formatTime": [Function],
        "formats": Object {},
        "formatters": Object {
          "getDateTimeFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralFormat": [Function],
          "getRelativeFormat": [Function],
        },
        "locale": "en",
        "messages": Object {},
        "now": [Function],
        "textComponent": "span",
      }
    }
    isFollowed={true}
    onClick={[Function]}
    pending={false}
    secondary={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        false,
        "Following",
      ],
      "className": "Follow",
      "onClick": [Function],
      "onMouseOut": [Function],
      "onMouseOver": [Function],
    },
    "ref": null,
    "rendered": Array [
      false,
      "Following",
    ],
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          false,
          "Following",
        ],
        "className": "Follow",
        "onClick": [Function],
        "onMouseOut": [Function],
        "onMouseOver": [Function],
      },
      "ref": null,
      "rendered": Array [
        false,
        "Following",
      ],
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
